cpp/src/hash/hashing.cu

```diff
@@ -77,84 +78,276 @@ class bitwise_partitioner
+                                   const partitioner_type the_partitioner,
+                                   size_type * row_partition_numbers,
+                                   size_type * block_partition_sizes,
+                                   size_type * global_partition_sizes)
+{
+  // Accumulate histogram of the size of each partition in shared memory
+  extern __shared__ size_type shared_partition_sizes[];
+
+  size_type row_number = threadIdx.x + blockIdx.x * blockDim.x;
+
+  // Initialize local histogram
+  size_type partition_number = threadIdx.x;
+  while(partition_number < num_partitions)
+  {
+    shared_partition_sizes[partition_number] = 0;
+    partition_number += blockDim.x;
+  }
+
+  __syncthreads();
+
+  // Compute the hash value for each row, store it to the array of hash values
+  // and compute the partition to which the hash value belongs and increment
+  // the shared memory counter for that partition
+  while( row_number < num_rows)
+  {
+    const hash_value_type row_hash_value = the_hasher(row_number);
+
+    const size_type partition_number = the_partitioner(row_hash_value);
+
+    row_partition_numbers[row_number] = partition_number;
+
+    atomicAdd(&(shared_partition_sizes[partition_number]), size_type(1));
+
+    row_number += blockDim.x * gridDim.x;
+  }
+
+  __syncthreads();
+
+  // Flush shared memory histogram to global memory
+  partition_number = threadIdx.x;
+  while(partition_number < num_partitions)
+  {
+    const size_type block_partition_size = shared_partition_sizes[partition_number];
+
+    // Update global size of each partition
+    atomicAdd(&global_partition_sizes[partition_number], block_partition_size);
+
+    // Record the size of this partition in this block
+    const size_type write_location = partition_number * gridDim.x + blockIdx.x;
+    block_partition_sizes[write_location] = block_partition_size;
+    partition_number += blockDim.x;
+  }
+}
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief  Given an array of partition numbers, computes the final output location
+   for each element in the output such that all rows with the same partition are 
+   contiguous in memory.
+ * 
+ * @param row_partition_numbers The array that records the partition number for each row
+ * @param num_rows The number of rows
+ * @param num_partitions THe number of partitions
+ * @param[out] block_partition_offsets Array that holds the offset of each partition for each thread block,
+ * i.e., { {block0 partition0 offset, block1 partition0 offset, ...}, 
+         {block0 partition1 offset, block1 partition1 offset, ...},
+         ...
+         {block0 partition(num_partitions-1) offset, block1 partition(num_partitions -1) offset, ...} }
+ */
+/* ----------------------------------------------------------------------------*/
+__global__ 
+void compute_row_output_locations(size_type * row_partition_numbers, 
```

Comment:
I suggest adding the restrict keyword to pointer kernel parameters:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#restrict
I've seen it improve performance of simple kernels by 5X (this one is far from simple, but it should still help).

Reply:
Good idea, but it doesn't seem to make a difference here. Here are the benchmark results with and without adding `__restrict__`
![Execution time for 32 cols per number of rows](https://user-images.githubusercontent.com/5580056/72655423-66e68280-3949-11ea-8402-8757e93b2134.png)

Reply:
:(
@trevorsm7 did you add the keyword to all pointer parameters?

Reply:
However, it didn't cause the unit tests to fail, so I suppose there's no harm in adding the keyword here anyway.

Reply:
Yes, I added it to all there pointer parameters of `compute_row_output_locations` and of `compute_row_partition_numbers`